tools/fix_mailmap.py

```diff
@@ -0,0 +1,573 @@
+import subprocess
+import sys
+from enum import IntEnum, unique
+from pathlib import Path
+from sys import stderr
+from unicodedata import normalize
+
+# Set up some globals for simplicity
+# pylint: disable=invalid-name, global-statement
+color_output = True
+quiet_mode = False  # Only report errors
+problem_found = False  # Problems requiring user intervention were detected
+changes_made = False  # .mailmap was updated
+mailmap_path = Path()  # We'll set it correctly later
+mailmap_body_offset = 0  # Offset from the .mailmap header into the body
+
+
+@unique
+class ExitCode(IntEnum):
+    """Program exit codes enumeration."""
+
+    NO_PROBLEM = 0
+    CHANGES_MADE = 1
+    MAILMAP_NOT_FOUND = 2
+    MAILMAP_FORMAT = 3
+    GIT_LOG = 4
+    MAILMAP_NEEDS_MANUAL_CHANGE = 5
+
+
+def mailmap_line(index: int) -> str:
+    """Format a string to refer to a mailmap line in console messages."""
+    return f"{mailmap_path}:{mailmap_body_offset+index}"
+
+
+def print_red(text: str) -> None:
+    """Print messages indicating need for user intervention."""
+    global problem_found
+    problem_found = True
+    if color_output:
+        text = "\n".join(f"\033[31m{line}\033[0m" for line in text.splitlines())
+    print(text, file=stderr)
+
+
+def print_yellow(text: str) -> None:
+    """Print messages indicating changes were made that the user should confirm."""
+    global changes_made
+    changes_made = True
+    if quiet_mode:
+        return
+    if color_output:
+        text = "\n".join(f"\033[33m{line}\033[0m" for line in text.splitlines())
+    print(text, file=stderr)
+
+
+def print_green(text: str) -> None:
+    """Print messages indicating no changes needed."""
+    if quiet_mode:
+        return
+    if color_output:
+        text = "\n".join(f"\033[32m{line}\033[0m" for line in text.splitlines())
+    print(text, file=stderr)
+
+
+def get_mailmap_path() -> Path:
+    """Use git to to find the relative path to the .mailmap file."""
+    cmd = ["git", "rev-parse", "--show-cdup"]
+    res = subprocess.run(cmd, capture_output=True, check=True, text=True)
+    return Path(res.stdout.strip()) / ".mailmap"
+
+
+def get_raw_authors() -> dict[str, set[tuple]]:
+    """Get the list of authors WITHOUT .mailmap remapping.
+
+    Returns a dict {lowercase_email: set((name, email))}, which will be used to check if a
+    mailmap line is necessary (matches some commit, and actually makes a change)
+    """
+    cmd = [
+        "git",
+        "log",
+        "--pretty=tformat:%an <%ae>%n%(trailers:key=co-authored-by,valueonly=true,unfold=true)",
+    ]
+    res = subprocess.run(cmd, capture_output=True, check=True, text=True)
+
+    logline = re.compile(r"([^<]+)\s+<(.*)>$")
+
+    emails_map: dict[str, set[tuple]] = {}
+    for line in res.stdout.splitlines():
+        match = logline.fullmatch(line.strip())
+        if match:
+            name, email = match.group(1, 2)
+            emails_map.setdefault(email.lower(), set()).add((name, email))
```

Comment:
Unimportant, but if you prefer, you can avoid the manual call to `setdefault` (and the construction of a set on every call) by making `emails_map = collections.defaultdict(set)`, then the `setdefault` line can just be `emails_map[email.lower()].add((name, email))`.